pkg/agent/clean/adunmigration/migrate.go

```diff
@@ -0,0 +1,506 @@
+			} else if errors.Is(err, LdapErrorNotFound{}) {
+				logrus.Debugf("[%v] user %v is GUID-based (%v) and the Active Directory server doesn't know about it. marking it as missing", identifyAdUserOperation, user.Name, guid)
+				knownGUIDMissingUnits[guid] = len(missingUsers)
+				missingUsers = append(missingUsers, missingUserWorkUnit{guid: guid, originalUser: user.DeepCopy()})
+			} else {
+				logrus.Debugf("[%v] user %v is GUID-based (%v) and the Active Directory server knows it by the Distinguished Name '%v'", identifyAdUserOperation, user.Name, guid, dn)
+				knownGUIDWorkUnits[guid] = len(usersToMigrate)
+				knownDnWorkUnits[dn] = len(usersToMigrate)
+				var emptyDuplicateList []*v3.User
+				usersToMigrate = append(usersToMigrate, migrateUserWorkUnit{guid: guid, distinguishedName: dn, originalUser: user.DeepCopy(), duplicateUsers: emptyDuplicateList})
+			}
+		}
+	}
+
+	if sharedLConn.isOpen {
+		sharedLConn.lConn.Close()
+	}
+
+	if len(usersToMigrate) == 0 {
+		logrus.Debugf("[%v] found 0 users in need of migration, exiting without checking for DN-based duplicates", identifyAdUserOperation)
+		return usersToMigrate, missingUsers, skippedUsers
+	}
+
+	// Now for the second pass, we need to identify DN-based users, and see if they are duplicates of any of the GUID
+	// users that we found in the first pass. We'll prefer the oldest user as the originalUser object, this will be
+	// the one we keep when we resolve duplicates later.
+	logrus.Debugf("[%v] locating any DN-based Active Directory users", identifyAdUserOperation)
+	for _, user := range users.Items {
+		if !isAdUser(&user) {
+			logrus.Debugf("[%v] user '%v' has no AD principals, skipping", identifyAdUserOperation, user.Name)
+			continue
+		}
+		principalID := adPrincipalID(&user)
+		logrus.Debugf("[%v] processing AD User '%v' with principal ID: '%v'", identifyAdUserOperation, user.Name, principalID)
+		if isGUID(principalID) {
+			logrus.Debugf("[%v] '%v' does not appear to be a DN-based principal ID, taking no action", identifyAdUserOperation, principalID)
+			continue
+		}
+		dn, err := getExternalID(principalID)
+		if err != nil {
+			logrus.Errorf("[%v] failed to extract DN from principal '%v', cannot process user: '%v'", identifyAdUserOperation, err, user.Name)
+			continue
+		}
+		if i, exists := knownDnWorkUnits[dn]; exists {
+			logrus.Debugf("[%v] user %v is DN-based (%v), and a duplicate of %v",
+				identifyAdUserOperation, user.Name, dn, usersToMigrate[i].originalUser.Name)
+			// Make sure the oldest duplicate user is selected as the original
+			if usersToMigrate[i].originalUser.CreationTimestamp.Time.After(user.CreationTimestamp.Time) {
+				usersToMigrate[i].duplicateUsers = append(usersToMigrate[i].duplicateUsers, usersToMigrate[i].originalUser)
+				usersToMigrate[i].originalUser = &user
+			} else {
+				usersToMigrate[i].duplicateUsers = append(usersToMigrate[i].duplicateUsers, &user)
+			}
```

Comment:
Just as a note, in the code above that looped over users a `user.DeepCopy()` is used and here it's a pointer with `&user`. Not sure if there is an impact of using the pointer here.

Reply:
I *think* we need a copy again here since `user` will keep changing and just storing the pointer to it will result things only getting mapped to the last user iterated over.  I'll confirm this with Nicholas.